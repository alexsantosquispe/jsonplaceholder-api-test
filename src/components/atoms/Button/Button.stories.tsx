import type { Meta, StoryObj } from '@storybook/react';

import { Button, ButtonProps } from './Button';
import { Wrapper } from '../../../utils/stories-utils';

const meta = {
  title: 'Atoms/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = (args: Partial<ButtonProps>) => {
  return (
    <div className="grid grid-cols-4 gap-4">
      <Wrapper className="flex-col">
        <Button label={args.label} />
        <span className="text-sm">Default</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label={args.label} isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary />
        <span className="text-sm">Secondary</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
    </div>
  );
};

Default.args = {
  label: 'Button',
  onClick: () => alert('Button clicked!')
};
