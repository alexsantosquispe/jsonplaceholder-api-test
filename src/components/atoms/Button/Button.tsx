import cn from 'clsx';
import { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

export interface ButtonProps {
  label?: string;
  type?: 'button' | 'submit';
  icon?: ReactNode;
  onClick?: () => void;
  isSecondary?: boolean;
  isDisabled?: boolean;
  className?: string;
  ariaLabel?: string;
}

export const Button = ({
  label,
  type = 'button',
  onClick,
  icon = null,
  isDisabled = false,
  isSecondary = false,
  className = '',
  ariaLabel
}: ButtonProps) => {
  return (
    <button
      role="button"
      aria-disabled={isDisabled}
      aria-label={ariaLabel}
      className={twMerge(
        'flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out',
        cn({
          'bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90':
            !isSecondary,
          'hover:bg-primary/80 dark:hover:bg-white/80': !isDisabled,
          'text-primary/80 hover:text-primary border border-black/10 bg-transparent shadow-xs hover:bg-gray-100 dark:border-white/15 dark:text-white dark:hover:bg-white/15':
            isSecondary,
          'text-primary/40 hover:text-primary/40 cursor-not-allowed border-black/5 hover:bg-transparent dark:border-white/10 dark:text-white/25 dark:hover:bg-transparent':
            isSecondary && isDisabled,
          'bg-primary/65 cursor-not-allowed dark:bg-white/25':
            !isSecondary && isDisabled
        }),
        className
      )}
      disabled={isDisabled}
      type={type}
      onClick={onClick}
    >
      {icon && <span className="size-4">{icon}</span>}
      {label && label}
    </button>
  );
};
