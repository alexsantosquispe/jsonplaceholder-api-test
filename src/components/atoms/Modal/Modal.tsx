import { twMerge } from 'tailwind-merge';

interface ModalProps {
  children: React.ReactNode;
  containerClassName?: string;
  onClose: () => void;
  onCancel?: () => void;
  onAccept?: () => void;
}

const HeaderModal = ({ onClose }: Pick<ModalProps, 'onClose'>) => {
  return (
    <div className="flex w-full justify-end gap-8 px-4 py-2">
      <button
        className="self-end text-gray-700 hover:cursor-pointer hover:text-black dark:text-white/70 dark:hover:text-white"
        onClick={onClose}
      >
        &#x2715;
      </button>
    </div>
  );
};

const FooterModal = ({
  onCancel,
  onAccept
}: Omit<ModalProps, 'children' | 'onClose'>) => {
  if (!onCancel || !onAccept) {
    return null;
  }

  return (
    <div className="flex w-full justify-end space-x-4 px-4 py-2">
      <button
        className="rounded-lg border border-gray-200 bg-white px-5 py-2.5 text-sm font-medium text-black hover:cursor-pointer hover:bg-gray-100 focus:ring-4 focus:ring-gray-100 focus:outline-none"
        onClick={onCancel}
      >
        Cancel
      </button>
      <button
        type="button"
        className="rounded-lg bg-black px-5 py-2.5 text-sm font-medium text-white hover:cursor-pointer hover:bg-black/80 focus:ring-4 focus:ring-gray-500 focus:outline-none"
        onClick={onAccept}
      >
        Accept
      </button>
    </div>
  );
};

export const Modal = ({
  children,
  onClose,
  onCancel,
  onAccept,
  containerClassName
}: ModalProps) => {
  return (
    <div className="fixed inset-0 z-50 flex items-center justify-center bg-black/80 backdrop-blur-sm dark:bg-black/15">
      <div className="m-4 flex h-8/10 w-full flex-col rounded-lg border border-gray-100 bg-white md:w-3/4 md:pb-8 lg:max-h-[35rem] lg:min-h-[40rem] lg:w-1/2 dark:border-white/10 dark:bg-black">
        <HeaderModal onClose={onClose} />

        <div
          className={twMerge(
            'flex w-full flex-col overflow-auto rounded-b-lg p-2 md:p-4',
            containerClassName
          )}
        >
          {children}
        </div>

        <FooterModal onCancel={onCancel} onAccept={onAccept} />
      </div>
    </div>
  );
};
