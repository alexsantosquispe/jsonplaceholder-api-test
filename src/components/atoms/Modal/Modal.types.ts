type ModalClassName = {
  container?: string;
  body?: string;
};

export interface ModalProps {
  children: React.ReactNode;
  onClose: () => void;
  title?: string;
  acceptLabel?: string;
  cancelLabel?: string;
  onAccept?: () => void;
  onCancel?: () => void;
  className?: ModalClassName;
}
