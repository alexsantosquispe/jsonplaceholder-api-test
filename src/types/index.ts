import { Control, FieldValues } from 'react-hook-form';

export interface SvgIconProps {
  className?: string;
}

export interface InputProps {
  label: string;
  name: string;
  control: Control<FieldValues>;
  placeholder?: string;
  type?: string;
  errorMessage?: string;
  isDisabled?: boolean;
  isRequired?: boolean;
}

export type TextAreaProps = Omit<InputProps, 'type'> & {
  rows?: number;
  cols?: number;
};
