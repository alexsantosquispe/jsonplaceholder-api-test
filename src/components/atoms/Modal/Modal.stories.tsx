import type { Meta, StoryObj } from '@storybook/react';
import { useState } from 'react';

import { Modal } from './Modal';
import { Button } from '../Button/Button';

const meta = {
  title: 'Atoms/Modal',
  component: Modal,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Modal>;

export default meta;

type Story = StoryObj<typeof Modal>;

export const Default: Story = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div className="flex h-screen w-full items-center justify-center">
      <Button label="Open modal" onClick={() => setIsOpen(true)} />
      {isOpen && (
        <Modal title="Modal test" onClose={() => setIsOpen(false)}>
          <div className="flex w-full py-4">
            Lorem, ipsum dolor sit amet consectetur adipisicing elit. Commodi
            deserunt dicta voluptates ipsam in modi dignissimos facilis nesciunt
            ab, voluptatibus incidunt autem tempora nostrum veritatis doloremque
            reprehenderit ipsum iste earum.
          </div>
        </Modal>
      )}
    </div>
  );
};

Default.args = {};
