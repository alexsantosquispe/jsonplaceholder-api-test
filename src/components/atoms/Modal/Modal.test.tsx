import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { useState } from 'react';

import { Modal } from './Modal';
import { ModalProps } from './Modal.types';
import { Button } from '../Button/Button';

const TestContentModal = (props: Omit<ModalProps, 'children' | 'onClose'>) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button label="Open modal" onClick={() => setIsOpen(true)} />
      {isOpen && (
        <Modal {...props} onClose={() => setIsOpen(false)}>
          <div className="h-80 w-full">This is a content for a modal</div>;
        </Modal>
      )}
    </div>
  );
};

describe('Modal', () => {
  const onClose = jest.fn();
  const props = {
    title: 'Modal title',
    acceptLabel: 'Accept',
    cancelLabel: 'Cancel'
  };

  it('should render the component correctly', () => {
    const component = render(
      <Modal {...props} onClose={onClose}>
        <span>test</span>
      </Modal>
    );

    expect(component).toMatchSnapshot();
  });

  describe('Behavior', () => {
    it('should open the modal', async () => {
      render(<TestContentModal {...props} />);

      const openModalButton = screen.getByText('Open modal');

      expect(openModalButton).toBeInTheDocument();

      await waitFor(() => {
        expect(screen.queryByText(props.title)).not.toBeInTheDocument();
      });

      fireEvent.click(openModalButton);

      await waitFor(() => {
        expect(screen.queryByText(props.title)).toBeInTheDocument();
      });
    });

    it('should close the modal', async () => {
      render(<TestContentModal {...props} />);

      fireEvent.click(screen.getByText('Open modal'));

      await waitFor(() => {
        expect(screen.queryByText(props.title)).toBeInTheDocument();
      });

      const closeButton = screen.getAllByRole('button')[1];

      fireEvent.click(closeButton);

      await waitFor(() => {
        expect(screen.queryByText(props.title)).not.toBeInTheDocument();
      });
    });
  });
});
