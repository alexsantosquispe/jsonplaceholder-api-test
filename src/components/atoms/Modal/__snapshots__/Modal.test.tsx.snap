// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed inset-0 z-50 flex items-end md:items-center md:justify-center bg-black/80 backdrop-blur-xs dark:bg-black/15"
      >
        <div
          class="flex w-full flex-col justify-between rounded-t-lg border md:rounded-lg md:w-[40rem] border-gray-100 bg-white dark:border-white/10 dark:bg-black"
        >
          <div
            class="flex w-full items-start justify-between gap-8 px-4 py-2"
          >
            <span
              class="py-1 text-lg font-medium"
            >
              Modal title
            </span>
            <button
              class="text-gray-700 hover:cursor-pointer hover:text-black dark:text-white/70 dark:hover:text-white"
            >
              ✕
            </button>
          </div>
          <div
            class="flex flex-col overflow-auto px-2 md:px-4"
          >
            <span>
              test
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed inset-0 z-50 flex items-end md:items-center md:justify-center bg-black/80 backdrop-blur-xs dark:bg-black/15"
    >
      <div
        class="flex w-full flex-col justify-between rounded-t-lg border md:rounded-lg md:w-[40rem] border-gray-100 bg-white dark:border-white/10 dark:bg-black"
      >
        <div
          class="flex w-full items-start justify-between gap-8 px-4 py-2"
        >
          <span
            class="py-1 text-lg font-medium"
          >
            Modal title
          </span>
          <button
            class="text-gray-700 hover:cursor-pointer hover:text-black dark:text-white/70 dark:hover:text-white"
          >
            ✕
          </button>
        </div>
        <div
          class="flex flex-col overflow-auto px-2 md:px-4"
        >
          <span>
            test
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
